engine/CSVParser.cc

```diff
@@ -447,26 +461,85 @@ namespace minsky
-  
-        v.setXVector(xVector);
-        if (!cminsky().checkMemAllocation(v.numElements()*sizeof(double)))
-          throw runtime_error("memory threshold exeeded");
-        // stash the data into vv tensorInit field
-        v.tensorInit.data.clear();
-        v.tensorInit.data.resize(v.numElements(), spec.missingValue);
-        auto dims=v.tensorInit.dims=v.dims();    
-        for (auto& i: tmpData)
-          {
-            size_t idx=0;
-            assert (dims.size()==i.first.size());
-            assert(dimLabels.size()==dims.size());
-            for (int j=dims.size()-1; j>=0; --j)
-              {
-                assert(dimLabels[j].count(i.first[j]));
-                idx = (idx*dims[j]) + dimLabels[j][i.first[j]];
-              }
-            v.tensorInit.data[idx]=i.second;
-          }
+                  
+        v.sparsityRatio =  static_cast<double>(1.0-static_cast<double>(tmpData.size())/v.numDenseElements()); 
+		
+		// for feature 47
+		if (v.sparsityRatio <= 0.5) 
+	    {
+          v.setXVector(xVector);          
+          if (!cminsky().checkMemAllocation(v.numDenseElements()*sizeof(double)))
+            throw runtime_error("memory threshold exceeded");
+		  v.index.clear();	            
+          // stash the data into vv tensorInit field
+          v.tensorInit.data.clear();
+          v.tensorInit.data.resize(v.numDenseElements(), spec.missingValue);  
+          auto dims=v.tensorInit.dims=v.dims();    
+          for (auto& i: tmpData)
+            {
+              size_t idx=0;
+              assert (dims.size()==i.first.size());
+              assert(dimLabels.size()==dims.size());
+              for (int j=dims.size()-1; j>=0; --j)
+                {
+                  assert(dimLabels[j].count(i.first[j]));
+                  idx = (idx*dims[j]) + dimLabels[j][i.first[j]];
+                }
+              v.tensorInit.data[idx]=i.second;  
+            }
+        }    
+        else 
+	    {	
+		  v.numSparseElements=tmpData.size();	
+          v.setXVector(xVector);          
+          if (!cminsky().checkMemAllocation(v.numSparseElements*sizeof(double)))
+            throw runtime_error("memory threshold exceeded");		  
+		  v.index.clear();			
+		  v.tensorInit.data.clear();	
+		  v.tensorInit.data.resize(v.numSparseElements, spec.missingValue); 	
+		  vector<double> tempData;
+          auto dims=v.tensorInit.dims=v.dims();    
+          for (auto& i: tmpData)
+            {
+              size_t idx=0;
+              assert (dims.size()==i.first.size());
+              assert(dimLabels.size()==dims.size());
+              for (int j=dims.size()-1; j>=0; --j)
+                {
+                  assert(dimLabels[j].count(i.first[j]));
+                  idx = (idx*dims[j]) + dimLabels[j][i.first[j]];
+                }
+               if (!isnan(i.second)) {
+                 v.index.push_back(idx);
+                 tempData.push_back(i.second);
+               }                   
```

Comment:
OK so far. there is some common code with the true part of the conditional that could be refactored, but lets leave that until we've got things working.

Reply:
Thanks. I directly load i.second into v.tensorInit.data now, since there is no reason not to use the same data structure for the sparse data. The only difference is that its size is v.numSparseElements (or better yet v.index.size()) instead of v.numDenseElements(). Its indices run from 0 to v.index.size()-1, whereas v.index contains the original flattened/serialized hypercube indices which, as I understand it, will be required anywhere in the code the hypercube axes (dimensions) are looped over and the data is used in that context.